### DISCUSSION TITLE
How to build the bin file of ESP32-S3 WIZNET5K to support Ethernet communication

### DISCUSSION THREAD

**Discussion by yfyungit** [micropython/micropython#16237] [ESP32] [answered]
I checked the following links and did not find instructions for building WIZNET5K or custom drivers
https://docs.micropython.org/en/latest/reference/manifest.html

Where can I find the relevant documentation?

I need to build ESP32-S3 16MB and 4MB flash versions to support WIZNET5K Ethernet communication using W5500

**Comment by bidrohini** [comment]
MicroPython's support for WIZNET5K Ethernet modules is controlled by the MICROPY_PY_NETWORK_WIZNET5K configuration option. You have to enable this.

**Reply by projectgus** [reply]
This is true for all MicroPython ports that supports Wiznet W5500, *except for esp32 port*. On esp32 port the driver is part of ESP-IDF (not MicroPython) and is enabled by default on most builds.

(This wasn't documented at the time this question was asked, but it's documented now. Details in other answer.)

**Comment by yfyungit** [comment]
![20241117171755](https://github.com/user-attachments/assets/dcc70f9c-ec74-4289-8923-a56c843da5a0)
@bidrohini After I enabled MICROPY_PY_NETWORK_WIZNET5K support, I got an error during the build process

**Reply by andrewleech** [reply]
The important part of the error is most likely above that block of text in the terminal. 

Also try a `make clean` after changing settings like that one, before trying the build again.

**Reply by yfyungit** [reply]
![image](https://github.com/user-attachments/assets/10fa47c2-1dcc-442b-9c36-406b63c4ef12)

I found this error, it seems to be caused by incorrect path. After I tried to change it to the correct path, another error occurred. The following figure shows the error after correcting the directory.

![image](https://github.com/user-attachments/assets/467885eb-eae9-422c-b595-ec66444e815e)

**Reply by yfyungit** [reply]
`In file included from /home/xkli/esp/micropython/lib/wiznet5k/Ethernet/wizchip_conf.h:173,
                 from /home/xkli/esp/micropython/extmod/network_wiznet5k.c:46:
/home/xkli/esp/micropython/lib/wiznet5k/Ethernet/W5500/w5500.h:218: warning: "MR" redefined
  218 | #define MR                 (_W5500_IO_BASE_ + (0x0000 << 8) + (WIZCHIP_CREG_BLOCK << 3))
      |
In file included from /home/xkli/esp/esp-idf/components/xtensa/include/xtensa/xtruntime.h:30,
                 from /home/xkli/esp/esp-idf/components/freertos/FreeRTOS-Kernel/portable/xtensa/include/freertos/portmacro.h:72,
                 from /home/xkli/esp/esp-idf/components/freertos/FreeRTOS-Kernel/include/freertos/portable.h:58,
                 from /home/xkli/esp/esp-idf/components/freertos/FreeRTOS-Kernel/include/freertos/FreeRTOS.h:70,
                 from /home/xkli/esp/micropython/ports/esp32/mpconfigport.h:11,
                 from /home/xkli/esp/micropython/py/mpconfig.h:91,
                 from /home/xkli/esp/micropython/py/mpstate.h:31,
                 from /home/xkli/esp/micropython/py/runtime.h:29,
                 from /home/xkli/esp/micropython/extmod/network_wiznet5k.c:32:
/home/xkli/esp/esp-idf/components/xtensa/esp32s3/include/xtensa/config/specreg.h:104: note: this is the location of the previous definition
  104 | #define MR              32
      |
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: In function 'wiz_cs_select':
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:134:5: error: implicit declaration of function 'mp_hal_pin_low'; did you mean 'mp_hal_pin_od_low'? [-Werror=implicit-function-declaration]
  134 |     mp_hal_pin_low(wiznet5k_obj.cs);
      |     ^~~~~~~~~~~~~~
      |     mp_hal_pin_od_low
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: In function 'wiz_cs_deselect':
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:138:5: error: implicit declaration of function 'mp_hal_pin_high'; did you mean 'mp_hal_pin_od_high'? [-Werror=implicit-function-declaration]
  138 |     mp_hal_pin_high(wiznet5k_obj.cs);
      |     ^~~~~~~~~~~~~~~
      |     mp_hal_pin_od_high
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: In function 'mpy_wiznet_yield':
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:144:5: error: implicit declaration of function 'MICROPY_THREAD_YIELD'; did you mean 'MICROPY_PY_THREAD_GIL'? [-Werror=implicit-function-declaration]
  144 |     MICROPY_THREAD_YIELD();
      |     ^~~~~~~~~~~~~~~~~~~~
      |     MICROPY_PY_THREAD_GIL
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: In function 'wiznet5k_init':
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:392:5: error: implicit declaration of function 'mod_network_register_nic' [-Werror=implicit-function-declaration]
  392 |     mod_network_register_nic(&wiznet5k_obj);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: At top level:
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:415:35: error: unknown type name 'mod_network_socket_obj_t'
  415 | STATIC int wiznet5k_socket_socket(mod_network_socket_obj_t *socket, int *_errno) {
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:459:35: error: unknown type name 'mod_network_socket_obj_t'
  459 | STATIC void wiznet5k_socket_close(mod_network_socket_obj_t *socket) {
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:467:33: error: unknown type name 'mod_network_socket_obj_t'
  467 | STATIC int wiznet5k_socket_bind(mod_network_socket_obj_t *socket, byte *ip, mp_uint_t port, int *_errno) {
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:483:35: error: unknown type name 'mod_network_socket_obj_t'
  483 | STATIC int wiznet5k_socket_listen(mod_network_socket_obj_t *socket, mp_int_t backlog, int *_errno) {
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:493:35: error: unknown type name 'mod_network_socket_obj_t'
  493 | STATIC int wiznet5k_socket_accept(mod_network_socket_obj_t *socket, mod_network_socket_obj_t *socket2, byte *ip, mp_uint_t *port, int *_errno) {
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:493:69: error: unknown type name 'mod_network_socket_obj_t'
  493 | STATIC int wiznet5k_socket_accept(mod_network_socket_obj_t *socket, mod_network_socket_obj_t *socket2, byte *ip, mp_uint_t *port, int *_errno) {
      |                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:528:36: error: unknown type name 'mod_network_socket_obj_t'
  528 | STATIC int wiznet5k_socket_connect(mod_network_socket_obj_t *socket, byte *ip, mp_uint_t port, int *_errno) {
      |                                    ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:549:39: error: unknown type name 'mod_network_socket_obj_t'
  549 | STATIC mp_uint_t wiznet5k_socket_send(mod_network_socket_obj_t *socket, const byte *buf, mp_uint_t len, int *_errno) {
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:563:39: error: unknown type name 'mod_network_socket_obj_t'
  563 | STATIC mp_uint_t wiznet5k_socket_recv(mod_network_socket_obj_t *socket, byte *buf, mp_uint_t len, int *_errno) {
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:577:41: error: unknown type name 'mod_network_socket_obj_t'
  577 | STATIC mp_uint_t wiznet5k_socket_sendto(mod_network_socket_obj_t *socket, const byte *buf, mp_uint_t len, byte *ip, mp_uint_t port, int *_errno) {
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:597:43: error: unknown type name 'mod_network_socket_obj_t'
  597 | STATIC mp_uint_t wiznet5k_socket_recvfrom(mod_network_socket_obj_t *socket, byte *buf, mp_uint_t len, byte *ip, mp_uint_t *port, int *_errno) {
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:611:39: error: unknown type name 'mod_network_socket_obj_t'
  611 | STATIC int wiznet5k_socket_setsockopt(mod_network_socket_obj_t *socket, mp_uint_t level, mp_uint_t opt, const void *optval, mp_uint_t optlen, int *_errno) {
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:617:39: error: unknown type name 'mod_network_socket_obj_t'
  617 | STATIC int wiznet5k_socket_settimeout(mod_network_socket_obj_t *socket, mp_uint_t timeout_ms, int *_errno) {
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:631:34: error: unknown type name 'mod_network_socket_obj_t'
  631 | STATIC int wiznet5k_socket_ioctl(mod_network_socket_obj_t *socket, mp_uint_t request, mp_uint_t arg, int *_errno) {
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: In function 'wiznet5k_make_new':
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:730:15: error: implicit declaration of function 'mp_hal_get_spi_obj'; did you mean 'mp_hal_get_pin_obj'? [-Werror=implicit-function-declaration]
  730 |         spi = mp_hal_get_spi_obj(args[0]);
      |               ^~~~~~~~~~~~~~~~~~
      |               mp_hal_get_pin_obj
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:730:13: warning: assignment to 'mp_obj_base_t *' {aka 'struct _mp_obj_base_t *'} from 'int' makes pointer from integer without a cast [-Wint-conversion]
  730 |         spi = mp_hal_get_spi_obj(args[0]);
      |             ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:745:48: error: 'mod_network_nic_type_wiznet5k' undeclared (first use in this function)
  745 |     wiznet5k_obj.base.type = (mp_obj_type_t *)&mod_network_nic_type_wiznet5k;
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:745:48: note: each undeclared identifier is reported only once for each function it appears in
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: In function 'wiznet5k_active':
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:842:21: error: implicit declaration of function 'mp_hal_get_mac' [-Werror=implicit-function-declaration]
  842 |                     mp_hal_get_mac(MP_HAL_MAC_ETH0, mac);
      |                     ^~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:842:36: error: 'MP_HAL_MAC_ETH0' undeclared (first use in this function)
  842 |                     mp_hal_get_mac(MP_HAL_MAC_ETH0, mac);
      |                                    ^~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c: At top level:
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1022:7: error: unknown type name 'mod_network_nic_protocol_t'
 1022 | const mod_network_nic_protocol_t mod_network_nic_protocol_wiznet = {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1023:5: error: field name not in record or union initializer
 1023 |     .gethostbyname = wiznet5k_gethostbyname,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1023:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1023:22: warning: initialization of 'int' from 'int (*)(void *, const char *, mp_uint_t,  uint8_t *)' {aka 'int (*)(void *, const char *, long unsigned int,  unsigned char *)'} makes integer from pointer without a cast [-Wint-conversion]
 1023 |     .gethostbyname = wiznet5k_gethostbyname,
      |                      ^~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1023:22: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1024:5: error: field name not in record or union initializer
 1024 |     .socket = wiznet5k_socket_socket,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1024:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1024:15: error: 'wiznet5k_socket_socket' undeclared here (not in a function); did you mean 'wiznet5k_locals_dict'?
 1024 |     .socket = wiznet5k_socket_socket,
      |               ^~~~~~~~~~~~~~~~~~~~~~
      |               wiznet5k_locals_dict
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1024:15: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1024:15: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1025:5: error: field name not in record or union initializer
 1025 |     .close = wiznet5k_socket_close,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1025:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1025:14: error: 'wiznet5k_socket_close' undeclared here (not in a function)
 1025 |     .close = wiznet5k_socket_close,
      |              ^~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1025:14: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1025:14: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1026:5: error: field name not in record or union initializer
 1026 |     .bind = wiznet5k_socket_bind,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1026:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1026:13: error: 'wiznet5k_socket_bind' undeclared here (not in a function)
 1026 |     .bind = wiznet5k_socket_bind,
      |             ^~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1026:13: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1026:13: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1027:5: error: field name not in record or union initializer
 1027 |     .listen = wiznet5k_socket_listen,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1027:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1027:15: error: 'wiznet5k_socket_listen' undeclared here (not in a function); did you mean 'wiznet5k_locals_dict'?
 1027 |     .listen = wiznet5k_socket_listen,
      |               ^~~~~~~~~~~~~~~~~~~~~~
      |               wiznet5k_locals_dict
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1027:15: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1027:15: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1028:5: error: field name not in record or union initializer
 1028 |     .accept = wiznet5k_socket_accept,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1028:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1028:15: error: 'wiznet5k_socket_accept' undeclared here (not in a function); did you mean 'wiznet5k_locals_dict'?
 1028 |     .accept = wiznet5k_socket_accept,
      |               ^~~~~~~~~~~~~~~~~~~~~~
      |               wiznet5k_locals_dict
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1028:15: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1028:15: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1029:5: error: field name not in record or union initializer
 1029 |     .connect = wiznet5k_socket_connect,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1029:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1029:16: error: 'wiznet5k_socket_connect' undeclared here (not in a function)
 1029 |     .connect = wiznet5k_socket_connect,
      |                ^~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1029:16: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1029:16: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1030:5: error: field name not in record or union initializer
 1030 |     .send = wiznet5k_socket_send,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1030:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1030:13: error: 'wiznet5k_socket_send' undeclared here (not in a function); did you mean 'wiznet5k_make_new'?
 1030 |     .send = wiznet5k_socket_send,
      |             ^~~~~~~~~~~~~~~~~~~~
      |             wiznet5k_make_new
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1030:13: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1030:13: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1031:5: error: field name not in record or union initializer
 1031 |     .recv = wiznet5k_socket_recv,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1031:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1031:13: error: 'wiznet5k_socket_recv' undeclared here (not in a function)
 1031 |     .recv = wiznet5k_socket_recv,
      |             ^~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1031:13: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1031:13: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1032:5: error: field name not in record or union initializer
 1032 |     .sendto = wiznet5k_socket_sendto,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1032:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1032:15: error: 'wiznet5k_socket_sendto' undeclared here (not in a function)
 1032 |     .sendto = wiznet5k_socket_sendto,
      |               ^~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1032:15: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1032:15: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1033:5: error: field name not in record or union initializer
 1033 |     .recvfrom = wiznet5k_socket_recvfrom,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1033:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1033:17: error: 'wiznet5k_socket_recvfrom' undeclared here (not in a function)
 1033 |     .recvfrom = wiznet5k_socket_recvfrom,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1033:17: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1033:17: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1034:5: error: field name not in record or union initializer
 1034 |     .setsockopt = wiznet5k_socket_setsockopt,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1034:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1034:19: error: 'wiznet5k_socket_setsockopt' undeclared here (not in a function)
 1034 |     .setsockopt = wiznet5k_socket_setsockopt,
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1034:19: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1034:19: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1035:5: error: field name not in record or union initializer
 1035 |     .settimeout = wiznet5k_socket_settimeout,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1035:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1035:19: error: 'wiznet5k_socket_settimeout' undeclared here (not in a function)
 1035 |     .settimeout = wiznet5k_socket_settimeout,
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1035:19: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1035:19: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1036:5: error: field name not in record or union initializer
 1036 |     .ioctl = wiznet5k_socket_ioctl,
      |     ^
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1036:5: note: (near initialization for 'mod_network_nic_protocol_wiznet')
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1036:14: error: 'wiznet5k_socket_ioctl' undeclared here (not in a function); did you mean 'wiznet5k_locals_dict'?
 1036 |     .ioctl = wiznet5k_socket_ioctl,
      |              ^~~~~~~~~~~~~~~~~~~~~
      |              wiznet5k_locals_dict
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1036:14: warning: excess elements in scalar initializer
/home/xkli/esp/micropython/extmod/network_wiznet5k.c:1036:14: note: (near initialization for 'mod_network_nic_protocol_wiznet')
cc1: some warnings being treated as errors
[1242/1293] Building C object esp-idf/main_esp32s3/CMakeFiles/...e/xkli/esp/micropython/extmod/nimble/modbluetooth_nimble.c.objninja: build stopped: subcommand failed.
HINT: Please check that the function name is correct. Also it is possible that you've forgot to import mp_hal_pin_low library(s) in header file or add the necessary REQURIES component. Try to add missing libraries to your project header file or check idf_component_register(REQUIRES ...) section in your component CmakeList.txt file. For more information run 'idf.py docs -sp api-guides/build-system.html'.
Also, please check if the function has been removed, renamed or replaced by an alternative function - refer to the migration guide for more information.
ninja failed with exit code 1, output of the command is in the /home/xkli/esp/micropython/ports/esp32/build-ESP32_GENERIC_S3_W5500/log/idf_py_stderr_output_75846 and /home/xkli/esp/micropython/ports/esp32/build-ESP32_GENERIC_S3_W5500/log/idf_py_stdout_output_75846
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [Makefile:66：all] 错误 1
`

**Reply by littlegreenbean33** [reply]
I just made it work for 1.24.1 / idf 5.2.2 just adding MICROPY_PY_NETWORK_WIZNET5K 1 to  ports/esp32/boards/ESP32_GENERIC_S3/sdkconfig.board

**Reply by yfyungit** [reply]
@littlegreenbean33 Thank you very much. This problem has troubled me for a long time. After following your instructions and building a new firmware, W5500 can work normally now.

**Reply by kevinf300p** [reply]
I'm a newbie trying to get WIZNET5K/W5500 to work on an ESP32_GENERIC_S3.  Been struggling to get WIZNET properly included.  when I add MICROPY_PY_NETWORK_WIZNET5K 1 to ports/esp32/boards/ESP32_GENERIC_S3/sdkconfig.board, I get...

/var/folders/8x/tyb23w3j08106x8twvsdft8d3qbsc8/T/kconfgen_tmp38tyg1kr:191: warning: ignoring malformed line 'MICROPY_PY_NETWORK_WIZNET5K 1'

**Reply by projectgus** [reply]
@kevinf300p The esp32 port uses a different driver for WIZNET5K to some other ports. Setting `MICROPY_PY_NETWORK_WIZNET5K` shouldn't be necessary, in particular it's not an sdkconfig option and I'd expect to see the same error you're seeing if adding to that file. I'm not sure exactly how the other two posters got this to work.

EDIT: I've put the key info into another post so I could mark it as an answer, see https://github.com/orgs/micropython/discussions/16237#discussioncomment-12608836 (one of the examples is even exactly your hardware setup, ESP32-S3 with W5500).

**Comment by yfyungit** [comment]
@littlegreenbean33 During the use, I still have a question: How to use WLAN and W5500 at the same time? For example, WLAN uses TCP to connect to server A, W5500 establishes a TCP server, and waits for the external TCP client to establish a connection. Both connections can be used normally. Can you share your experience or provide a simple demo code for reference?

**Reply by littlegreenbean33** [reply]
you need to bind the socket to the address of the specific [interface](https://docs.micropython.org/en/v1.24.0/library/socket.html#socket.socket.bind). That way you listen on the specific address associated to the specific interface. 

Because Micropython/Esp32 do not do (by default) forwarding between the  network interfaces, the client and server will only be able to connect to devices in the same network. (i.e. either wifi or LAN)

You need to take of a asynchronous nature of your case. Assuming reading writing to a socket, while at the same time having a server listening on the second interface. Threads,

**Comment by projectgus** [comment, marked as answer]
The esp32 port already supports Wiznet W5500, but it uses a different method to configure it compared to other ports. See newly added docs here: https://docs.micropython.org/en/latest/esp32/quickref.html#spi-ethernet-interface

(These docs weren't there at the time of this question.)

**Reply by yfyungit** [reply]
Thank you very much, with relevant documents this will help many people
